notion-core/src/main/scala/com/ambiata/notion/core/LocationIO.scala

```diff
@@ -276,17 +201,79 @@ case class LocationIO(configuration: Configuration, s3Client: AmazonS3Client) {
+  def reduceBytes[T](location: Location, reducer: BytesReducer[T]): LocationIO[T] =
     streamBytes[reducer.S](location, reducer.init)((bytes, s) => reducer.reduce(bytes, s)).map(s => reducer.finalise(s))
 
   /** get the last line of a file */
-  def tail(location: Location, numberOfLines: Int): RIO[List[String]] =
+  def tail(location: Location, numberOfLines: Int): LocationIO[List[String]] =
     reduceLinesUTF8(location, LineReducer.tail(numberOfLines))
 
   /** get the last line of a file */
-  def lastLine(location: Location): RIO[Option[String]] =
+  def lastLine(location: Location): LocationIO[Option[String]] =
     tail(location, numberOfLines = 1).map(_.lastOption)
+
+  def writeUtf8Lines(location: Location, lines: List[String]): LocationIO[Unit] =
+    writeUtf8(location, Lists.prepareForFile(lines))
+
+  def writeUtf8(location: Location, string: String): LocationIO[Unit] =
+    withLocationContext(location) {
+      case LocalLocationContext(p)      => fromRIO(p.write(string)).void
+      case HdfsLocationContext(conf, p) => fromHdfs(conf, p.write(string)).void
+      case S3LocationContext(client, p) => fromS3(client, for {
+        e <- p.exists
+        _ <- S3Action.when(e, S3Action.fail(s"Can not overwrite location ${location}"))
+        _ <- p.put(string)
+      } yield ())
+    }
+
+  def writeBytes(location: Location, bytes: Array[Byte]): LocationIO[Unit] =
+    withLocationContext(location) {
+      case LocalLocationContext(p)      => fromRIO(p.writeBytes(bytes)).void
+      case HdfsLocationContext(conf, p) => fromHdfs(conf, p.writeBytes(bytes)).void
+      case S3LocationContext(client, p) => fromS3(client, for {
+        e <- p.exists
+        _ <- S3Action.when(e, S3Action.fail(s"Can not overwrite location ${location}"))
+        _ <- p.putBytes(bytes)
+      } yield ())
+    }
+
+  def deleteAll(location: Location): LocationIO[Unit] =
+    withLocationContext(location) {
+      case LocalLocationContext(p)      => fromRIO(p.delete)
+      case HdfsLocationContext(conf, p) => fromHdfs(conf, p.delete)
+      case S3LocationContext(client, p) => fromS3(client, p.delete)
+    }
+
+  def delete(location: Location): LocationIO[Unit] =
+    withLocationContext(location) {
+      case LocalLocationContext(p)      => fromRIO(p.determineFile.flatMap(_.delete))
+      case HdfsLocationContext(conf, p) => fromHdfs(conf, p.determineFile.flatMap(_.delete))
+      case S3LocationContext(client, p) => fromS3(client, p.determineAddress.flatMap(_.delete))
+    }
+
+  /**
+   * copy a file by simply piping lines from one location to the other
+   *
+   * if overwrite is false and the location file already exists, return an Error
+   */
+  def copyFile(from: Location, to: Location, overwrite: Boolean): LocationIO[Unit] = for {
+    e <- exists(to)
+    _ <- when(e) {
+      if(overwrite) delete(to)
+      else LocationIO(_ => RIO.fail[LocationIOResult[Unit]](s"Can not overwrite location ${to}"))
+    }
+    _ <- pipe(from, to)
```

Comment:
This is pretty inefficient, mainly for s3 where saws provides a copy which doesn't go through the box (this will)